lib/installedtest.pm

```diff
@@ -288,12 +289,54 @@ sub maybe_unlock_screen {
 
 sub save_and_upload_log {
     my ($self, $cmd, $file, $args) = @_;
-    script_run("$cmd > $file", $args->{timeout});
-    my $ret = upload_logs($file) unless $args->{noupload};
+    script_run("$cmd > $file", timeout=>$args->{timeout});
+    my $ret = upload_logs(
+        $file,
+        timeout=>$args->{timeout},
+        failok=>$args->{failok}
+    ) unless $args->{noupload};
     save_screenshot if $args->{screenshot};
+    return undef if ($args->{failok} and !-e "ulogs/$ret");
     return $ret;
 }
 
+sub upload_packages_versions {
+    my ($self, %args) = @_;
+
+    my $all_packages = "";
+
+    # log package versions
+    my $fname = $self->save_and_upload_log('rpm -qa', 'dom0-packages.txt',
+                                           {failok=>$args{failok}});
+    if ($fname) {
+        my $packages = path('ulogs', $fname)->slurp;
+        $packages = join("\nDom0: ", sort split(/\n/, $packages));
+        $all_packages .= "Dom0: " . $packages . "\n";
+    } else {
+        $all_packages .= "Dom0: failed\n";
+    };
+
+    my $templates = script_output('qvm-ls --raw-data --fields name,klass');
+    foreach (sort split /\n/, $templates) {
+        next unless /Template/;
+        s/\|.*//;
+        $fname = $self->save_and_upload_log("qvm-run --no-gui -ap $_ 'rpm -qa; dpkg -l; pacman -Q; true'",
+                "template-$_-packages.txt",
+                {timeout =>90, failok=>$args{failok}});
+        if ($fname) {
+            my $packages = path('ulogs', $fname)->slurp;
+            $packages = join("\n$_: ", sort split(/\n/, $packages));
+            $all_packages .= $_ . ": " . $packages . "\n";
+        } else {
+            $all_packages .= $_ . ": failed\n";
```

Comment:
Just upload should have 10s timeout now, but I guess collecting the info doesn't make much sense if it's going to fail anyway...
Since this `last` is effectively under `failok` condition anyway, yes, I'm okay with this.